providers/enchant_voikko.c

```diff
@@ -108,14 +105,10 @@ static int
 voikko_provider_dictionary_exists (struct str_enchant_provider * me _GL_UNUSED_PARAMETER,
                                    const char *const tag)
 {
-	/* Only Finnish is supported */
-	if (strncmp(tag, "fi", 3) != 0)
-		return 0;
-
```

Comment:
This made me scratch my head. 

The branch: https://github.com/albbas/enchant/commits/fix_voikko_test is a sketchy pass at making these tests pass.

This commit: https://github.com/albbas/enchant/commit/5a91921bf9752935dbd2d5f63a839b51494586e4 made the voikko provider tests pass. The next commit: https://github.com/albbas/enchant/commit/991cfc5ccc2b83029aa2bac248fc033cf6c1673d made the other provider tests pass.

The crash in the tests seems to be connected to the fact that some providers use ''enchant_provider_set_error'' to set the the enchant provider error message, while others do not.

Reply:
I would rather, at least for this pull request, have a fix that does not change the test or any other provider.

I see that the `dictionary_exists` method is not documented, in common with other provider methods, so it’s a matter of debate what it should do.

The test `ProviderDictionaryExists_ProviderDoesNotHave_ProviderDoesNotSetError` suggests that providers are expected not to set an error just because a requested dictionary does not exist. Hence, I think we should stick to this behaviour for now.

Hence, I think the right solution is to change your patch to the voikko back-end to respect this.

Does that sound OK?

Reply:
That sounds perfectly reasonable, I'll try to fix the patch that way.

Reply:
This all looks good now, thanks! Please could you squash the commits into a single commit? Hopefully also that will get us a clean Travis run (there seems to be a failure owing to some VM problem in one of the latest macOS builds).

Reply:
The squash has been done, hopefully everything works as it should.

Reply:
Thanks very much. Now that I can see a single commit, I think there's still one problem: the provider _should_ call `enchant_provider_set_error` if there is some error other than the dictionary not being found. (I guess that is why there used to be the hard-wired check for the language tag.)

The problem is that `voikkoInit` does not give an easy way to tell these cases apart (except perhaps by checking the error message, but that is undocumented). A possibly expensive but correct way (with a slight but probably acceptable race condition) would be to check whether the language is supported by calling `voikko_list_dicts` first, and afterwards assume that a `NULL` return from `voikkoInit` indicates an error.

Reply:
I tried to solve that problem, too.
I had to shuffle the code around a bit to do the changes, so I did it three commits to make it easier to see the changes.

I see a possible problem with the code in **voikko_provider_dictionary_exists**. If the **tag** is eg. fi_FI, it will not match the language provided by voikko, **fi**.

I'll find out how voikko can return tags like fi_FI.

Reply:
@albbas, the lack of e.g. "fi_FI" should not be a problem: Enchant itself tries the base language, here "fi", if the variant is not found.

Please can you squash the commits into a single commit? I think the changes are simple enough they're actually easier to review that way.

Good to see all the tests are now passing!